pkg/completion/completion.go

```diff
@@ -103,38 +118,43 @@ type Completion struct {
 	completed chan struct{}
 
 	// callback is called when Complete is called the first time.
-	callback func()
+	callback func(err error)
+
+	// err is the error the completion completed with
+	Error error
```

Comment:
I would not export this field, since all accesses to this field must be done while locking using `lock`. So I would rename this `Error` -> `Err`, and define a new `Error()` method that returns this field while holding `lock`.

Reply:
This field can be ready without locking after calling Complete(), or after the Completed() channel is closed. Added a comment to that effect.

Reply:
No. It typically can't be read without locking if it's written by another goroutine. Cf. https://golang.org/ref/mem#tmp_2
> When multiple goroutines access a shared variable v, they must use synchronization events to establish happens-before conditions that ensure reads observe the desired writes.

Reply:
Made the member non-exported by renaming to `err`. Added a new `Err()` method mirroring the method of the same name of `Context`.

Reply:
Added a commit to make it clear and consistent how WaitGroup's Wait() prioritizes error values for return. Also addressed the concern of accessing completion's 'err' without locking.